### ISSUE TITLE
Feature request: Restart application after update

### ISSUE THREAD

**Issue by peters** [Squirrel/OldSquirrelForWindows#130]
It would be nice to restart the application after a successfully applied update. 

Shimmer.Client.Restart()

See https://github.com/github/Shimmer/issues/118#issuecomment-24181867 https://github.com/github/Shimmer/issues/118#issuecomment-24184128

**Comment by shiftkey** [comment]
Yes, the recommended ways to do this are rather [mediocre](http://stackoverflow.com/questions/4773632/how-do-i-restart-a-wpf-application).

If someone wants to propose a patch for this I'm open to it.

**Comment by peters** [comment]
This solution works well for my projects: https://gist.github.com/peters/6518385

**Comment by shiftkey** [comment]
The limitation with that approach is it requires code _outside_ of Shimmer (i.e. the entry point to your application) to work properly...

**Comment by peters** [comment]
Sure, but in one of my applications i'm using a global mutex (kiosk software). It's not possible to guarantee that this mutex will be released before the application is restarted using suggested approaches@stackoverflow.

You could create a forward method where you execute Application.Current.Shutdown in the context of a Dispatcher derived from the Assembly you wish to restart. 

``` cs
Action<Dispatcher> forward = (dp) => dp.Invoke(Application.Current.Shutdown);
```

![code_monkey_desktop_by_ispaz247](https://f.cloud.github.com/assets/869/1119807/1188be36-1a86-11e3-9b6e-1d3807eff15e.jpg)

**Comment by shiftkey** [comment]
![image](https://f.cloud.github.com/assets/359239/1119819/e88ebf16-1a86-11e3-8a46-c9a1782b1001.png)

**Comment by allenpetersen** [comment]
I'm thinking about moving the Shimmer parts of my app into a separate AppLauncher.EXE that will either start the main app if there were no updates or start the newly installed main app.

**Comment by shiftkey** [comment]
@allenpetersen a bootstrapper-esque library you could wrap the WPF application in might help solve this problem (like you do with enforcing a single-instance application using a Mutex). 

Let me know if anything interesting spins out of that which might fit with the rest of the Shimmer project.

**Comment by peters** [comment]
But: This is probably not relevant to most people but has anyone tought about a rollback solution if everything goes to shit during an upgrade (on my part i have a demo staging area, deploying updates to multiple machines before it ever reaches the end user). But if shit hits the fan, it would be ideal that the application does a form of rollback and starts the old version.

**Comment by shiftkey** [comment]
> This is probably not relevant to most people but has anyone tought about a rollback solution if everything goes to shit during an upgrade

Given you have the old packages handy, I can't see any technical limitations about this.

I guess it just depends on: 
- what you define as a bad upgrade (that it fails to complete, or that the current version has issues)
- whether a rollback can be as simple as installing version N-1
- how rollbacks should behave with things like AppSetup (more complexity headaches?)

**Comment by anaisbetts** [comment]
> But if shit hits the fan, it would be ideal that the application does a form of rollback and starts the old version.

Shimmer was explicitly designed such that there is actually a pretty small window where users get left with a bad version. We switch the shortcuts at the very, very end of the process - if anything fails before then, nothing happens and the old version gets run instead.

**Comment by anaisbetts** [comment]
> It would be nice to restart the application after a successfully applied update.

We can certainly do this, it'd take some Clever Hax™ but it's definitely not overly difficult. Basically:
1. Grab the currently running EXE's path by using `Assembly.GetEntryAssembly()`
2. Grab out the command line from [Environment](http://msdn.microsoft.com/en-us/library/system.environment.commandline.aspx)
3. `String.Replace` the old EXE with the path to the new one (if that EXE has been removed and replaced with a differently named one, this now becomes Tricky™)
4. `Process.Start`, then `Environment.FailFast` ourselves to get out ASAP.

**Comment by peters** [comment]
Discovered another use case for this yesterday: Modifying HKEY_CURRENT_USER registry keys

**Comment by patrickjohncollins** [comment]
I do not see how the solution proposed by @peters (https://github.com/Squirrel/Squirrel.Windows/issues/130#issuecomment-24207839) is any help, for after an installation the new release is in a different sub folder.
Before : app-1.0.5266.22823\app.exe
After : app-1.0.5266.23129\app.exe
What we need is to run the app in the new folder, although I'm not sure of the best way to achieve this.
